File: displayer/common/streamer/VideoMaker.py
#!/usr/bin/env python
# coding: utf-8

import cv2
import datetime


class VideoMaker:
    def __init__(self, cap, filename=None, fps=None, width=None, height=None):

        if not cap or not cap.isOpened():
            raise ValueError('Invalid video capture object')
        else:
            self.cap = cap

        # Save the filename
        if filename is None:
            self.filename = datetime.datetime.now().strftime('%Y-%m-%d_%H-%M-%S')
        else:
            self.filename = filename

        # If no FPS is provided, get it from the camera
        if not fps:
            self.fps = int(cap.get(cv2.CAP_PROP_FPS))
        else:
            self.fps = fps

        # If no width is provided, get it from the camera
        if not width:
            self.width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        else:
            self.width = width

        # If no height is provided, get it from the camera
        if not height:
            self.height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        else:
            self.height = height

        # Initialize the VideoWriter
        self.video = self.initialize_video_writer()

    def initialize_video_writer(self):
        # Attempt to initialize VideoWriter with MP4 codec
        video = cv2.VideoWriter(self.filename + '.mp4', cv2.VideoWriter_fourcc(*'mp4v'),
                                self.fps,
                                (self.width, self.height))
        if video.isOpened():
            return video

        # Fallback to AVI codec
        self.filename += '.avi'
        video = cv2.VideoWriter(self.filename, cv2.VideoWriter_fourcc(*'XVID'),
                                self.fps,
                                (self.width, self.height))
        if video.isOpened():
            return video

        raise ValueError('Failed to initialize video writer with both MP4 and AVI codecs')

    def add_frame(self, frame):
        # Get the size of the frame
        frame_width = frame.shape[1]
        frame_height = frame.shape[0]

        # Check if the frame size is the same as the video size
        if frame_width != self.width or frame_height != self.height:
            # Resize the frame to match the video dimensions
            frame = cv2.resize(frame, (self.width, self.height))

        # Write the frame to the video
        self.video.write(frame)

    def release(self):
        print('Closing video file')
        self.video.release()


if __name__ == "__main__":
    # Use VideoCapture to open the camera
    cap = cv2.VideoCapture(0)

    if not cap.isOpened():
        raise ValueError('Invalid video capture object')

    # Get the camera properties
    width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
    height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
    fps = 20

    # Use VideoMaker, specifying keyword arguments explicitly
    video_maker = VideoMaker(cap, fps=fps, width=width, height=height)

    while True:
        ret, frame = cap.read()
        if not ret:
            break

        video_maker.add_frame(frame)
        cv2.imshow('frame', frame)

        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    # Release the VideoMaker and the camera
    video_maker.release()
    cap.release()
    cv2.destroyAllWindows()

File: jetson/common/stream/VideoMaker.cpp
//
// Created by orlando on 9/25/24.
//

#include "VideoMaker.h"

VideoMaker::VideoMaker(cv::VideoCapture& cap, const std::string& name, int fps, int width, int height)
        : cap(cap) {
    if (!cap.isOpened()) {
        throw std::invalid_argument("Invalid video capture object");
    }

    // Check if the name has an extension
    size_t dot_pos = name.find_last_of('.');
    if (dot_pos != std::string::npos) {
        // Remove the extension
        this->name = name.substr(0, dot_pos);
    } else {
        this->name = name;
    }

    this->name = this->name + "_" + getCurrentTimestamp();

    // Get or set FPS
    this->fps = (fps == 0) ? static_cast<int>(cap.get(cv::CAP_PROP_FPS)) : fps;

    // Get or set width and height
    this->width = (width == 0) ? static_cast<int>(cap.get(cv::CAP_PROP_FRAME_WIDTH)) : width;
    this->height = (height == 0) ? static_cast<int>(cap.get(cv::CAP_PROP_FRAME_HEIGHT)) : height;

    // Initialize VideoWriter
    initializeVideoWriter();
}

void VideoMaker::addFrame(const cv::Mat& frame) {
    if (frame.empty()) {
        std::cerr << "Empty frame detected, skipping." << std::endl;
        return;
    }

    if (frame.cols != this->width || frame.rows != this->height) {
        // Resize the frame to match the video size instead of throwing an exception
        cv::Mat resized_frame;
        cv::resize(frame, resized_frame, cv::Size(this->width, this->height));
        this->video.write(resized_frame);
    } else {
        // Write frame to the video
        this->video.write(frame);
    }
}

void VideoMaker::release() {
    std::cout << "Closing video file: " << this->name + ".mp4" << std::endl;
    this->video.release();
}

std::string VideoMaker::getCurrentTimestamp() {
    auto now = std::chrono::system_clock::now();
    std::time_t now_time = std::chrono::system_clock::to_time_t(now);
    std::tm tm_now = *std::localtime(&now_time);

    std::ostringstream oss;
    oss << std::put_time(&tm_now, "%Y-%m-%d_%H-%M-%S");
    return oss.str();
}

void VideoMaker::initializeVideoWriter() {
    // Try to open video writer with MP4 codec
    std::string video_name = this->name + ".mp4";
    bool initialized = video.open(video_name, cv::VideoWriter::fourcc('m', 'p', '4', 'v'), this->fps, cv::Size(this->width, this->height));

    // If MP4 fails, fallback to AVI codec
    if (!initialized) {
        std::cerr << "Failed to initialize MP4, falling back to AVI" << std::endl;
        video_name = this->name + ".avi";
        initialized = video.open(video_name, cv::VideoWriter::fourcc('X', 'V', 'I', 'D'), this->fps, cv::Size(this->width, this->height));
    }

    // If both codecs fail, throw an exception
    if (!initialized) {
        throw std::runtime_error("Failed to initialize video writer with both MP4 and AVI codecs");
    }

    std::cout << "Video recording started: " << video_name << std::endl;
}